src/channel/mod.rs

```diff
@@ -1,14 +1,89 @@
+pub mod channel_states;
+
+pub mod fs_verifier_channel;
+pub mod fs_prover_channel;
+
 use ark_ff::Field;
+use channel_states::ChannelStates;
+use std::convert::{AsMut, AsRef};
+use crate::hashutil::TempHashContainer;
 
 #[allow(dead_code)]
-trait Channel {
+trait Channel: AsRef<ChannelStates> + AsMut<ChannelStates> {
     type Field: Field;
 
-    fn recv_felts(&mut self, n: usize) -> Result<Vec<Self::Field>, anyhow::Error>;
+    fn recv_felem(&mut self, felem: Self::Field) -> Result<Self::Field, anyhow::Error>;
 
     fn recv_bytes(&mut self, n: usize) -> Result<Vec<u8>, anyhow::Error>;
 
     fn random_number(&mut self, bound: u64) -> u64;
 
     fn random_field(&mut self) -> Self::Field;
+
+    /// Only relevant for non-interactive channels. Changes the channel seed to a "safer" seed.
+    ///
+    /// This function guarantees that randomness fetched from the channel after calling this function
+    /// and before sending data from the prover to the channel, is "safe": A malicious
+    /// prover will have to perform 2^security_bits operations for each attempt to randomize the fetched
+    /// randomness.
+    ///
+    /// Increases the amount of work a malicious prover needs to perform, in order to fake a proof.
+    #[allow(unused_variables)]
+    fn apply_proof_of_work(&mut self, security_bits: usize) -> Result<(), anyhow::Error> {
```

Comment:
I would consider adding a trait which inherits from the channel with this method:


```rust
trait FSChannel: Channel {
    fn apply_proof_of_work(&mut self, security_bits: usize) -> Result<(), anyhow::Error>;
}
```

That way you don't need to have implementations what just error :)